File: Helpers/DueDatePresets.cs
namespace Hatch.Helpers;

internal static class DueDatePresets
{
    public static DateTime GetToday(DateTime today) => today;

    public static DateTime GetTomorrow(DateTime today) => today.AddDays(1);

    // Saturday of the current week. Sunday snaps to next Saturday (weekend already passed).
    // Formula: ((6 - dayOfWeek + 7) % 7) gives 0 on Sat, 6 on Sun, correct for all others.
    public static DateTime GetThisWeekend(DateTime today)
    {
        int daysUntilSaturday = ((int)DayOfWeek.Saturday - (int)today.DayOfWeek + 7) % 7;
        return today.AddDays(daysUntilSaturday);
    }

    // Monday of next week; if today is already Monday, return next Monday (+7).
    public static DateTime GetNextWeek(DateTime today)
    {
        int daysUntilMonday = ((int)DayOfWeek.Monday - (int)today.DayOfWeek + 7) % 7;
        return today.AddDays(daysUntilMonday == 0 ? 7 : daysUntilMonday);
    }
}

File: mobile/shared/src/commonMain/kotlin/dev/hatch/sync/DueDatePresets.kt
package dev.hatch.sync

import kotlinx.datetime.DateTimeUnit
import kotlinx.datetime.LocalDate
import kotlinx.datetime.isoDayNumber
import kotlinx.datetime.plus

// Kotlin transcription of windows/Helpers/DueDatePresets.cs.
object DueDatePresets {

    fun today(today: LocalDate): LocalDate = today

    fun tomorrow(today: LocalDate): LocalDate = today.plus(1, DateTimeUnit.DAY)

    // Saturday of the current week. Sunday snaps to next Saturday (weekend already passed).
    fun thisWeekend(today: LocalDate): LocalDate =
        today.plus((SATURDAY - weekDayNumber(today) + 7) % 7, DateTimeUnit.DAY)

    // Monday of next week; if today is already Monday, return next Monday (+7).
    fun nextWeek(today: LocalDate): LocalDate {
        val daysUntilMonday = (MONDAY - weekDayNumber(today) + 7) % 7
        return today.plus(if (daysUntilMonday == 0) 7 else daysUntilMonday, DateTimeUnit.DAY)
    }

    // C# numbers the week from Sunday = 0; ISO numbers it from Monday = 1 with Sunday = 7.
    // Taking the ISO number modulo 7 lands exactly on the C# value, which is what lets the
    // two formulas above stay readable as transcriptions rather than re-derivations.
    private fun weekDayNumber(date: LocalDate) = date.dayOfWeek.isoDayNumber % 7

    private const val MONDAY = 1
    private const val SATURDAY = 6
}
